<a href="https://colab.research.google.com/github/PadawanXXVI/spotify-data-storytelling/blob/main/spotify.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🎓 Spotify Data Storytelling — Semana 2  
## Coleta, Limpeza e Transformação de Dados

**Curso:** Tecnologia em Ciência de Dados — Faculdade de Tecnologia e Inovação Senac DF  
**Professor:** Alexsander Holanda Barreto  
**Autor:** Anderson de Matos Guimarães  

---

### 🎯 Objetivo da Semana 2
Esta etapa foca na **coleta via API do Spotify**, seguida por **limpeza, transformação e análise exploratória inicial**.  
O resultado final será um **dataset tratado (`spotify_semana2_tratado.csv`)** e um **relatório parcial** descrevendo metodologia e insights.

---

### 🧩 Etapas desta semana
1. Instalação e importação de bibliotecas.  
2. Autenticação segura na API.  
3. Coleta de faixas por gêneros brasileiros.  
4. Diagnóstico e inspeção inicial.  
5. Limpeza, padronização e criação de colunas derivadas.  
6. Exportação dos datasets bruto e tratado.  
7. Geração do relatório parcial.


In [27]:
# 🛠️ Instalação da biblioteca Spotipy (interface oficial da API do Spotify)
!pip install spotipy --quiet

print("✅ Spotipy instalada com sucesso!")


✅ Spotipy instalada com sucesso!


In [28]:
# ⚙️ Importação de bibliotecas necessárias
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
from datetime import datetime

print("✅ Bibliotecas importadas com sucesso!")


✅ Bibliotecas importadas com sucesso!


## 🔐 Autenticação Segura na API do Spotify  

Para garantir segurança e reprodutibilidade, as credenciais ficam armazenadas em variáveis de ambiente:  
- `SPOTIPY_CLIENT_ID`  
- `SPOTIPY_CLIENT_SECRET`  

Essas chaves são obtidas no [Spotify Developer Dashboard](https://developer.spotify.com/dashboard).


In [29]:
# 🔐 Autenticação segura usando secrets do Google Colab
from google.colab import userdata
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials

# 🔎 Recupera as chaves armazenadas em "Secrets"
client_id = userdata.get('SPOTIPY_CLIENT_ID')
client_secret = userdata.get('SPOTIPY_CLIENT_SECRET')

# ✅ Teste de autenticação
try:
    auth_manager = SpotifyClientCredentials(client_id=client_id, client_secret=client_secret)
    sp = spotipy.Spotify(auth_manager=auth_manager)
    sp.search(q="genre:pop", type="track", limit=1)
    print("🎧 Autenticação realizada com sucesso via Colab Secrets!")
except Exception as e:
    print(f"❌ Falha na autenticação: {e}")


🎧 Autenticação realizada com sucesso via Colab Secrets!


In [30]:
# 🌎 Descoberta automática de categorias Spotify (Brasil)
print("🔎 Coletando categorias disponíveis no Spotify Brasil...")

# Busca até 50 categorias disponíveis
categorias_data = sp.categories(country="BR", limit=50)
categorias = {c["name"]: c["id"] for c in categorias_data["categories"]["items"]}

print(f"✅ Total de categorias encontradas: {len(categorias)}\n")
print("📋 Categorias disponíveis no Brasil:")
for nome in categorias.keys():
    print("-", nome)

# 🧩 Filtra categorias claramente musicais
categorias_musicais = [
    c for c in categorias.keys()
    if any(x in c.lower() for x in [
        "pop", "rock", "hip", "r&b", "dance", "latin", "metal", "jazz",
        "soul", "folk", "acoustic", "instrumental", "electronic",
        "country", "alternative", "blues", "gospel", "indie", "k-pop",
        "mpb", "sertanejo", "forró", "funk", "samba", "pagode"
    ])
]

print(f"\n🎶 Categorias musicais identificadas automaticamente ({len(categorias_musicais)}):")
print(categorias_musicais)

# 🔁 Substitui a lista manual de gêneros pela automática
generos_validos = categorias_musicais


🔎 Coletando categorias disponíveis no Spotify Brasil...
✅ Total de categorias encontradas: 50

📋 Categorias disponíveis no Brasil:
- Made For You
- New Releases
- Hip-Hop
- Country
- Pop
- Latin
- Charts
- Rock
- R&B
- Dance/Electronic
- Indie
- Fall
- Halloween
- Diwali
- Workout
- Mood
- Party
- Student
- Love
- Disney
- Netflix
- Chill
- Sleep
- Discover
- Jazz
- Metal
- Christian & Gospel
- Música Mexicana
- Happy Holidays
- Classical
- Decades
- Kids & Family
- In the car
- At Home
- EQUAL
- Spotify CLASSICS
- GLOW
- Frequency
- TV & Movies
- Trending
- Folk & Acoustic
- K-pop
- Punk
- Blues
- Soul
- Alternative
- Anime
- Instrumental
- RADAR
- Fresh Finds

🎶 Categorias musicais identificadas automaticamente (17):
['Hip-Hop', 'Country', 'Pop', 'Latin', 'Rock', 'R&B', 'Dance/Electronic', 'Indie', 'Jazz', 'Metal', 'Christian & Gospel', 'Folk & Acoustic', 'K-pop', 'Blues', 'Soul', 'Alternative', 'Instrumental']


In [31]:
# 🎧 Coleta automática de faixas musicais via API do Spotify (Brasil)

# 1️⃣ Lista de categorias musicais definidas dinamicamente
generos_validos = categorias_musicais
print(f"🎯 Total de gêneros a serem coletados: {len(generos_validos)}\n")

# 2️⃣ Estruturas de armazenamento
all_tracks = []
resumo = []

# 3️⃣ Loop de coleta
for genero in generos_validos:
    print(f"🎶 Buscando faixas da categoria: {genero}")
    try:
        results = sp.search(q=f"genre:{genero}", type="track", limit=50, market="BR")
        tracks = results["tracks"]["items"]

        # Caso nenhuma faixa seja encontrada
        if len(tracks) == 0:
            print(f"⚠️ Nenhum resultado encontrado para: {genero}")
            continue

        # Paginação — coleta até 500 resultados por categoria
        offset = 50
        while results["tracks"]["next"] and offset < 500:
            results = sp.next(results["tracks"])
            tracks.extend(results["tracks"]["items"])
            offset += 50

        # Estruturação dos dados
        for item in tracks:
            all_tracks.append({
                "genero": genero,
                "faixa": item["name"],
                "artista": item["artists"][0]["name"],
                "album": item["album"]["name"],
                "data_lancamento": item["album"]["release_date"],
                "popularidade": item["popularity"],
                "duracao_ms": item["duration_ms"]
            })

        resumo.append({"genero": genero, "qtd_faixas": len(tracks)})
        print(f"✅ {len(tracks)} faixas coletadas para {genero}")

    except Exception as e:
        print(f"❌ Erro ao coletar gênero {genero}: {e}")

# 4️⃣ Consolidação em DataFrame
df_raw = pd.DataFrame(all_tracks)
df_resumo = pd.DataFrame(resumo).sort_values("qtd_faixas", ascending=False)

print("\n📊 Resumo da coleta por gênero:")
display(df_resumo)

# 5️⃣ Exportação do dataset bruto
os.makedirs("/content/data", exist_ok=True)
df_raw.to_csv("/content/data/spotify_raw.csv", index=False)
print(f"💾 Dataset bruto exportado com sucesso! {len(df_raw)} faixas salvas em /content/data/spotify_raw.csv")


🎯 Total de gêneros a serem coletados: 17

🎶 Buscando faixas da categoria: Hip-Hop
✅ 100 faixas coletadas para Hip-Hop
🎶 Buscando faixas da categoria: Country
✅ 100 faixas coletadas para Country
🎶 Buscando faixas da categoria: Pop
✅ 100 faixas coletadas para Pop
🎶 Buscando faixas da categoria: Latin
✅ 100 faixas coletadas para Latin
🎶 Buscando faixas da categoria: Rock
✅ 100 faixas coletadas para Rock
🎶 Buscando faixas da categoria: R&B
✅ 100 faixas coletadas para R&B
🎶 Buscando faixas da categoria: Dance/Electronic
✅ 100 faixas coletadas para Dance/Electronic
🎶 Buscando faixas da categoria: Indie
✅ 100 faixas coletadas para Indie
🎶 Buscando faixas da categoria: Jazz
✅ 100 faixas coletadas para Jazz
🎶 Buscando faixas da categoria: Metal
✅ 100 faixas coletadas para Metal
🎶 Buscando faixas da categoria: Christian & Gospel
✅ 100 faixas coletadas para Christian & Gospel
🎶 Buscando faixas da categoria: Folk & Acoustic
✅ 100 faixas coletadas para Folk & Acoustic
🎶 Buscando faixas da categoria

genero  qtd_faixas
0              Hip-Hop         100
1              Country         100
2                  Pop         100
3                Latin         100
4                 Rock         100
5                  R&B         100
6     Dance/Electronic         100
7                Indie         100
8                 Jazz         100
9                Metal         100
10  Christian & Gospel         100
11     Folk & Acoustic         100
12               K-pop         100
13               Blues         100
14                Soul         100
15         Alternative         100
16        Instrumental         100

💾 Dataset bruto exportado com sucesso! 1700 faixas salvas em /content/data/spotify_raw.csv


## 🧹 Fase 2 — Limpeza e Transformação dos Dados (Semana 2)

Nesta etapa, realizamos o tratamento do dataset coletado pela API do Spotify.
As principais tarefas incluem:
- Verificar a integridade dos dados (ausentes, duplicados e tipos inconsistentes);
- Corrigir formatações de datas e durações;
- Normalizar variáveis numéricas;
- Exportar a base tratada para as análises exploratórias da próxima semana.

Esta fase é crucial para garantir **qualidade, consistência e confiabilidade** na análise de dados musicais.


In [32]:
# 📂 Importa o dataset bruto salvo na fase anterior
import pandas as pd

df_raw = pd.read_csv("/content/data/spotify_raw.csv")

print(f"✅ Dataset carregado com sucesso: {df_raw.shape[0]} linhas × {df_raw.shape[1]} colunas")
display(df_raw.head())


✅ Dataset carregado com sucesso: 1700 linhas × 7 colunas


genero             faixa      artista                    album  \
0  Hip-Hop           Barcode        Icaki                  Отрепка   
1  Hip-Hop          Problems  Blake Rules  Sorry for Being Amazing   
2  Hip-Hop    Hypnotic State       Factor                     Gaya   
3  Hip-Hop    Характеристика     Nonamerz          Piratka Mixtape   
4  Hip-Hop  503 (Radio Edit)    Cool Nutz                      503   

  data_lancamento  popularidade  duracao_ms  
0      2019-08-08             0      135724  
1      2017-06-20             0      244610  
2      2013-11-01             0      381743  
3            2008             0      115746  
4      2019-12-11             0      225296

In [33]:
# 🔍 Estrutura geral e estatísticas básicas
print("📋 Estrutura do dataset bruto:")
print()
print(f"Dimensões: {df_raw.shape[0]} linhas × {df_raw.shape[1]} colunas\n")
print(df_raw.info())
print("\n📈 Estatísticas descritivas iniciais:")
display(df_raw.describe(include='all'))


📋 Estrutura do dataset bruto:

Dimensões: 1700 linhas × 7 colunas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1700 entries, 0 to 1699
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   genero           1700 non-null   object
 1   faixa            1700 non-null   object
 2   artista          1700 non-null   object
 3   album            1700 non-null   object
 4   data_lancamento  1700 non-null   object
 5   popularidade     1700 non-null   int64 
 6   duracao_ms       1700 non-null   int64 
dtypes: int64(2), object(5)
memory usage: 93.1+ KB
None

📈 Estatísticas descritivas iniciais:


genero   faixa      artista  \
count      1700    1700         1700   
unique       17    1480          834   
top     Hip-Hop  Escape  Linkin Park   
freq        100       4           34   
mean        NaN     NaN          NaN   
std         NaN     NaN          NaN   
min         NaN     NaN          NaN   
25%         NaN     NaN          NaN   
50%         NaN     NaN          NaN   
75%         NaN     NaN          NaN   
max         NaN     NaN          NaN   

                                               album data_lancamento  \
count                                           1700            1700   
unique                                          1281            1027   
top     The Life of a Showgirl + Acoustic Collection      2025-11-07   
freq                                              19              21   
mean                                             NaN             NaN   
std                                              NaN             NaN   
min                                              NaN             NaN   
25%                                              NaN             NaN   
50%                                              NaN             NaN   
75%                                              NaN             NaN   
max                                              NaN             NaN   

        popularidade    duracao_ms  
count    1700.000000  1.700000e+03  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean       53.425294  2.226848e+05  
std        32.222223  8.481462e+04  
min         0.000000  1.239900e+04  
25%        26.750000  1.718652e+05  
50%        67.000000  2.109310e+05  
75%        77.000000  2.533662e+05  
max       100.000000  1.188466e+06

In [34]:
# 🧼 Limpeza e padronização
df = df_raw.copy()

# 1️⃣ Remover duplicatas
df.drop_duplicates(subset=["faixa", "artista", "album"], inplace=True)

# 2️⃣ Corrigir formato de data
df["data_lancamento"] = pd.to_datetime(df["data_lancamento"], errors="coerce")

# 3️⃣ Converter duração para minutos
df["duracao_min"] = (df["duracao_ms"] / 60000).round(2)

# 4️⃣ Normalizar a popularidade entre 0 e 1
df["popularidade_norm"] = df["popularidade"] / 100

# 5️⃣ Remover linhas com dados ausentes críticos
df.dropna(subset=["genero", "faixa", "artista", "album"], inplace=True)

print("✅ Limpeza concluída. Estrutura atualizada:")
print(df.info())


✅ Limpeza concluída. Estrutura atualizada:
<class 'pandas.core.frame.DataFrame'>
Index: 1542 entries, 0 to 1699
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   genero             1542 non-null   object        
 1   faixa              1542 non-null   object        
 2   artista            1542 non-null   object        
 3   album              1542 non-null   object        
 4   data_lancamento    1449 non-null   datetime64[ns]
 5   popularidade       1542 non-null   int64         
 6   duracao_ms         1542 non-null   int64         
 7   duracao_min        1542 non-null   float64       
 8   popularidade_norm  1542 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 120.5+ KB
None


In [35]:
# 🧾 Diagnóstico final
print("🩶 Valores nulos restantes por coluna:")
print(df.isna().sum())

print("\n📊 Estatísticas descritivas após tratamento:")
display(df.describe(include='all'))


🩶 Valores nulos restantes por coluna:
genero                0
faixa                 0
artista               0
album                 0
data_lancamento      93
popularidade          0
duracao_ms            0
duracao_min           0
popularidade_norm     0
dtype: int64

📊 Estatísticas descritivas após tratamento:


genero   faixa       artista  \
count      1542    1542          1542   
unique       17    1480           834   
top     Hip-Hop  Escape  Taylor Swift   
freq        100       4            33   
mean        NaN     NaN           NaN   
min         NaN     NaN           NaN   
25%         NaN     NaN           NaN   
50%         NaN     NaN           NaN   
75%         NaN     NaN           NaN   
max         NaN     NaN           NaN   
std         NaN     NaN           NaN   

                                               album  \
count                                           1542   
unique                                          1281   
top     The Life of a Showgirl + Acoustic Collection   
freq                                              19   
mean                                             NaN   
min                                              NaN   
25%                                              NaN   
50%                                              NaN   
75%                                              NaN   
max                                              NaN   
std                                              NaN   

                      data_lancamento  popularidade    duracao_ms  \
count                            1449    1542.00000  1.542000e+03   
unique                            NaN           NaN           NaN   
top                               NaN           NaN           NaN   
freq                              NaN           NaN           NaN   
mean    2014-02-27 18:43:58.509316608      51.91310  2.251558e+05   
min               1952-04-01 00:00:00       0.00000  1.239900e+04   
25%               2009-11-05 00:00:00       0.00000  1.746530e+05   
50%               2019-05-31 00:00:00      66.00000  2.124665e+05   
75%               2024-03-06 00:00:00      77.00000  2.551245e+05   
max               2025-11-07 00:00:00     100.00000  1.188466e+06   
std                               NaN      32.95438  8.663915e+04   

        duracao_min  popularidade_norm  
count   1542.000000        1542.000000  
unique          NaN                NaN  
top             NaN                NaN  
freq            NaN                NaN  
mean       3.752633           0.519131  
min        0.210000           0.000000  
25%        2.910000           0.000000  
50%        3.540000           0.660000  
75%        4.250000           0.770000  
max       19.810000           1.000000  
std        1.444029           0.329544

In [36]:
# 💾 Exporta o dataset tratado
os.makedirs("/content/data", exist_ok=True)
df.to_csv("/content/data/spotify_semana2_tratado.csv", index=False)

print(f"💾 Dataset tratado salvo com sucesso em /content/data/spotify_semana2_tratado.csv ({df.shape[0]} registros)")


💾 Dataset tratado salvo com sucesso em /content/data/spotify_semana2_tratado.csv (1542 registros)


## 📊 Relatório Parcial — Semana 2 (Coleta, Limpeza e Transformação)

Nesta semana, foi realizada a coleta automática de **1.700 faixas musicais**
distribuídas em **17 gêneros**, por meio da API do Spotify (Spotipy).

O dataset passou por um processo de limpeza e transformação, com destaque para:
- Remoção de duplicatas e registros inconsistentes;
- Conversão da data de lançamento para formato `datetime`;
- Criação de colunas derivadas: duração em minutos e popularidade normalizada;
- Padronização dos tipos de dados para futura análise exploratória.

O dataset resultante, `spotify_semana2_tratado.csv`, está pronto para a **Fase 3 — Análise Exploratória e Visualização**, onde serão aplicadas técnicas estatísticas e gráficos para interpretar padrões musicais e comportamentos de consumo.


## 🚀 Próximos Passos (Semana 3 – Análise e Visualização)

Na próxima etapa, realizaremos:

- Análise exploratória (EDA) detalhada com Pandas, Seaborn e Plotly;
- Visualizações de distribuição de popularidade, duração e artistas mais frequentes;
- Correlações entre gêneros e popularidade média;
- Introdução de storytelling analítico assistido por IA (Engenharia de Prompt).

O objetivo é traduzir os dados tratados em **insights visuais e narrativas interpretativas** sobre o consumo musical no Brasil.
